src/scores.jl

```diff
@@ -490,65 +490,85 @@ function flattened_score_names(chain_data::ChainScoreData)::Array{String, 1}
-        end
-    end
+     open(filename, "w") do f
+         if isempty(score_names) # by default, export all scores from chain
+             score_names = flattened_score_names(chain_data)
+         end
+
+         column_names = String[] # colum names of the CSV
+         nested_keys = Dict{String, String}() # map score key to nested key, if any
+
+         num_districts = nothing
+         for score_name in score_names
+             score, nested_key = get_score_by_name(chain_data, score_name)
+             if score isa DistrictScore || score isa DistrictAggregate
+                 if isnothing(num_districts)
+                     initial_vals = chain_data.step_values[1]
+                     if !isnothing(nested_key)
+                         initial_vals = initial_vals[nested_key]
+                     end
+                     num_districts = length(initial_vals[score_name])
+                 end
+                 # add new column for each district
+                 district_columns = ["$(score_name)_$(i)" for i in 1:num_districts]
+                 append!(column_names, district_columns)
+             elseif score isa CompositeScore
+                 throw(ArgumentError("Cannot automatically save a CompositeScore to CSV. You may save a CompositeScore's child score."))
+             elseif score isa PlanScore
+                 push!(column_names, score_name)
+             end
+             if !isnothing(nested_key)
+                 nested_keys[score_name] = nested_key
+             end
+         end
+         # write column names
+         colname_row = hcat(column_names...) # turn into 1xn array
+         writedlm(f, colname_row, ',') # write intial row of column names
+
+         # iterate through all steps of chain
+         query = ChainScoreQuery(score_names, chain_data)
+         for step_values in query
+             row_values = []
+             for key in score_names
+                 if haskey(nested_keys, key)
+                     nested_key = nested_keys[key]
+                     append!(row_values, step_values[nested_key][key])
+                 else
+                     append!(row_values, step_values[key])
+                 end
+             end
+             # write one row for every state of the chain
+             writedlm(f, hcat(row_values...), ',')
+         end
+     end
+ end
+
+
+ function save_scores_to_json(filename::String,
+                              chain_data::ChainScoreData,
+                              score_names::Array{String,1}=String[])
+     """ Save the `scores` in a JSON file named `filename`.
+     """
+     open(filename, "w") do f
+         if isempty(score_names) # by default, export all scores from chain
+             score_names = flattened_score_names(chain_data)
+         end
+
+         # iterate through all steps of chain
+         query = ChainScoreQuery(score_names, chain_data)
+         first_entry = true
+         println(f, "[")
```

Comment:
Why are we manually writing brackets and commas? Would it be better to construct an array and then serialize the whole thing?

Reply:
I think the point here is that constructing an array and serializing the whole thing would take up a lot of main memory? And the reason for writing row-by-row is to avoid ever having to hold in main memory the entire of the score values throughout the chain

Reply:
Adding a comment to address this!